### PR TITLE
[SPARK-38570][SQL] Incorrect DynamicPartitionPruning caused by Literal

### PR DESCRIPTION
**Pull request by mcdull-zhang**
### What changes were proposed in this pull request?

The return value of Literal.references is an empty AttributeSet, so Literal is mistaken for a partition column.

For example, the sql in the test case will generate such a physical plan when the adaptive is closed: 
```text
*(4) Project [store_id#5281, date_id#5283, state_province#5292]
+- *(4) BroadcastHashJoin [store_id#5281], [store_id#5291], Inner, BuildRight, false
   :- Union
   :  :- *(1) Project [4 AS store_id#5281, date_id#5283]
   :  :  +- *(1) Filter ((isnotnull(date_id#5283) AND (date_id#5283 >= 1300)) AND dynamicpruningexpression(4 IN dynamicpruning#5300))
   :  :     :  +- ReusedSubquery SubqueryBroadcast dynamicpruning#5300, 0, [store_id#5291], [id=#336]
   :  :     +- *(1) ColumnarToRow
   :  :        +- FileScan parquet default.fact_sk[date_id#5283,store_id#5286] Batched: true, DataFilters: [isnotnull(date_id#5283), (date_id#5283 >= 1300)], Format: Parquet, Location: CatalogFileIndex(1 paths)[file:/Users/dongdongzhang/code/study/spark/spark-warehouse/org.apache.s..., PartitionFilters: [dynamicpruningexpression(4 IN dynamicpruning#5300)], PushedFilters: [IsNotNull(date_id), GreaterThanOrEqual(date_id,1300)], ReadSchema: struct<date_id:int>
   :  :              +- SubqueryBroadcast dynamicpruning#5300, 0, [store_id#5291], [id=#336]
   :  :                 +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, true] as bigint)),false), [id=#335]
   :  :                    +- *(1) Project [store_id#5291, state_province#5292]
   :  :                       +- *(1) Filter (((isnotnull(country#5293) AND (country#5293 = US)) AND ((store_id#5291 <=> 4) OR (store_id#5291 <=> 5))) AND isnotnull(store_id#5291))
   :  :                          +- *(1) ColumnarToRow
   :  :                             +- FileScan parquet default.dim_store[store_id#5291,state_province#5292,country#5293] Batched: true, DataFilters: [isnotnull(country#5293), (country#5293 = US), ((store_id#5291 <=> 4) OR (store_id#5291 <=> 5)), ..., Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/dongdongzhang/code/study/spark/spark-warehouse/org.apache...., PartitionFilters: [], PushedFilters: [IsNotNull(country), EqualTo(country,US), Or(EqualNullSafe(store_id,4),EqualNullSafe(store_id,5))..., ReadSchema: struct<store_id:int,state_province:string,country:string>
   :  +- *(2) Project [5 AS store_id#5282, date_id#5287]
   :     +- *(2) Filter ((isnotnull(date_id#5287) AND (date_id#5287 <= 1000)) AND dynamicpruningexpression(5 IN dynamicpruning#5300))
   :        :  +- ReusedSubquery SubqueryBroadcast dynamicpruning#5300, 0, [store_id#5291], [id=#336]
   :        +- *(2) ColumnarToRow
   :           +- FileScan parquet default.fact_stats[date_id#5287,store_id#5290] Batched: true, DataFilters: [isnotnull(date_id#5287), (date_id#5287 <= 1000)], Format: Parquet, Location: CatalogFileIndex(1 paths)[file:/Users/dongdongzhang/code/study/spark/spark-warehouse/org.apache.s..., PartitionFilters: [dynamicpruningexpression(5 IN dynamicpruning#5300)], PushedFilters: [IsNotNull(date_id), LessThanOrEqual(date_id,1000)], ReadSchema: struct<date_id:int>
   :                 +- ReusedSubquery SubqueryBroadcast dynamicpruning#5300, 0, [store_id#5291], [id=#336]
   +- ReusedExchange [store_id#5291, state_province#5292], BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, true] as bigint)),false), [id=#335]
```

after this pr:
```text
*(4) Project [store_id#5281, date_id#5283, state_province#5292]
+- *(4) BroadcastHashJoin [store_id#5281], [store_id#5291], Inner, BuildRight, false
   :- Union
   :  :- *(1) Project [4 AS store_id#5281, date_id#5283]
   :  :  +- *(1) Filter (isnotnull(date_id#5283) AND (date_id#5283 >= 1300))
   :  :     +- *(1) ColumnarToRow
   :  :        +- FileScan parquet default.fact_sk[date_id#5283,store_id#5286] Batched: true, DataFilters: [isnotnull(date_id#5283), (date_id#5283 >= 1300)], Format: Parquet, Location: CatalogFileIndex(1 paths)[file:/Users/dongdongzhang/code/study/spark/spark-warehouse/org.apache.s..., PartitionFilters: [], PushedFilters: [IsNotNull(date_id), GreaterThanOrEqual(date_id,1300)], ReadSchema: struct<date_id:int>
   :  +- *(2) Project [5 AS store_id#5282, date_id#5287]
   :     +- *(2) Filter (isnotnull(date_id#5287) AND (date_id#5287 <= 1000))
   :        +- *(2) ColumnarToRow
   :           +- FileScan parquet default.fact_stats[date_id#5287,store_id#5290] Batched: true, DataFilters: [isnotnull(date_id#5287), (date_id#5287 <= 1000)], Format: Parquet, Location: CatalogFileIndex(1 paths)[file:/Users/dongdongzhang/code/study/spark/spark-warehouse/org.apache.s..., PartitionFilters: [], PushedFilters: [IsNotNull(date_id), LessThanOrEqual(date_id,1000)], ReadSchema: struct<date_id:int>
   +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, true] as bigint)),false), [id=#326]
      +- *(3) Project [store_id#5291, state_province#5292]
         +- *(3) Filter (((isnotnull(country#5293) AND (country#5293 = US)) AND ((store_id#5291 <=> 4) OR (store_id#5291 <=> 5))) AND isnotnull(store_id#5291))
            +- *(3) ColumnarToRow
               +- FileScan parquet default.dim_store[store_id#5291,state_province#5292,country#5293] Batched: true, DataFilters: [isnotnull(country#5293), (country#5293 = US), ((store_id#5291 <=> 4) OR (store_id#5291 <=> 5)), ..., Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/dongdongzhang/code/study/spark/spark-warehouse/org.apache...., PartitionFilters: [], PushedFilters: [IsNotNull(country), EqualTo(country,US), Or(EqualNullSafe(store_id,4),EqualNullSafe(store_id,5))..., ReadSchema: struct<store_id:int,state_province:string,country:string>
```

### Why are the changes needed?
Execution performance improvement 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added unit test

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala [hunk 1]
```diff
@@ -65,7 +65,7 @@ object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper {
           case fs: HadoopFsRelation =>
             val partitionColumns = AttributeSet(
               l.resolve(fs.partitionSchema, fs.sparkSession.sessionState.analyzer.resolver))
-            if (resExp.references.subsetOf(partitionColumns)) {
+            if (!resExp.references.isEmpty && resExp.references.subsetOf(partitionColumns)) {
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala:68]: `resExp.references.nonEmpty`?

Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala:68]: Does the `DataSourceV2ScanRelation` also need this check?

Comment (mcdull-zhang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala]: HadoopFsRelation, HiveTableRelation, DataSourceV2ScanRelation all need this check, and now I filter the literal at the beginning.

Is such code reasonable? Or should I handle it in findExpressionAndTrackLineageDown? Or should it be handled in three case statements?

Thank you for your review

**Comment by mcdull-zhang** [comment]
friendly ping @cloud-fan

**Comment by mcdull-zhang** [comment]
friendly ping @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala (hunk 1 above)
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala]: Could we fix `PredicateHelper.findExpressionAndTrackLineageDown`?

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala [hunk 2]
```diff
@@ -1528,6 +1528,55 @@ abstract class DynamicPartitionPruningSuiteBase
       }
     }
   }
+
+  test("SPARK-38570: Fix incorrect DynamicPartitionPruning caused by Literal") {
+    withSQLConf(SQLConf.DYNAMIC_PARTITION_PRUNING_ENABLED.key -> "true") {
+      withTable("fact1", "fact2", "dim") {
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:1534]: We already have lots of built-in tables. Cloud we change the test query to:
```sql
SELECT f.store_id,
       f.date_id,
       s.state_province
FROM   (SELECT 4 AS store_id,
               date_id,
               product_id
        FROM   fact_sk
        WHERE  date_id >= 1300
        UNION ALL
        SELECT 5 AS store_id,
               date_id,
               product_id
        FROM   fact_stats
        WHERE  date_id <= 1000) f
       JOIN dim_store s
         ON f.store_id = s.store_id
WHERE  s.country = 'US'
```

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:1534]: +1

Comment (mcdull-zhang) [reply sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala]: So embarrassing, I wanted to use these built-in tables before, but I didn't see table fact_stats.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 3]
```diff
@@ -128,6 +128,9 @@ trait PredicateHelper extends AliasHelper with Logging {
   def findExpressionAndTrackLineageDown(
       exp: Expression,
       plan: LogicalPlan): Option[(Expression, LogicalPlan)] = {
+    if (exp.references.isEmpty) {
+      return None
+    }
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:133]: no biggie

```suggestion
    if (exp.references.isEmpty) return None
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala (hunk 1 above)
Comment (mcdull-zhang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala]: Is this ok now?

**Comment by HyukjinKwon** [comment]
cc @maryannxue too

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala [hunk 4]
```diff
@@ -1528,6 +1528,34 @@ abstract class DynamicPartitionPruningSuiteBase
...
+          |               date_id,
+          |               product_id
+          |      FROM   fact_stats
+          |      WHERE  date_id <= 1000) f
+          |JOIN dim_store s
+          |ON f.store_id = s.store_id
+          |WHERE s.country = 'US'
+          |""".stripMargin)
+
+      checkPartitionPruningPredicate(df, false, withBroadcast = false)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:1555]:
```suggestion
      checkPartitionPruningPredicate(df, withSubquery = false, withBroadcast = false)
```

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala (hunk 1 above)
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala]: LGTM.

**Review by wangyum — APPROVED** [review]

**Review by weixiuli — APPROVED** [review]
LGTM +1

**Comment by cloud-fan** [comment]
@mcdull-zhang can you re-trigger the test?

**Comment by wangyum** [comment]
Merged to master and branch-3.3.

**Comment by wangyum** [comment]
@mcdull-zhang Could you backport this to branch-3.2.

**Comment by mcdull-zhang** [comment]
@wangyum I have newly submitted a pr for 3.2. 

thanks again all.

**Pull request closed without merging**